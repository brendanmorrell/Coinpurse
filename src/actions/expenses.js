// PRIOR TO ADDING FIREBASE (NON-ASYN VERSION)
// component calls action generator
// action generator returns an object
// component dispatches object
// redux store changes

// ASYNCHRONOUS VERSION (AFTER FIREBASE INTEGRATION)
// component calls action generator
// action generator returns FUNCTION (not an object)
// component dispatches FUNCTION (not object)
// funciton runs (has the ability to dispatch other actions and do whatever it wants)
// (by defual, redux does not allow you to dispatch functions. need to set up a piece of redux middleware, 'redux-thunk', and grab 'applyMiddleware' from redux on the file you use to configure your store')


import database from '../firebase/firebase';


// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense,
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => { // this is the function we will be sending now that we have redux-thunk as opposed to just sending the object we originally sent
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0,
    } = expenseData;

    const expense = { description, note, amount, createdAt };

    return database.ref('expenses').push(expense).then((ref) => { // the return is for promise chaining in the test file
      dispatch(addExpense({
        id: ref.key,
        ...expense,
      }));
    });
  };
};

// REMOVE_EXPENSE
export const removeExpense = id => ({
  type: 'REMOVE_EXPENSE',
  id,
});
// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates,
});
