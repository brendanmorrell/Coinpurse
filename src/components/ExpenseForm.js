import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';// this is an airbnb available module that gives a calendar. other versions also available

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      errorDescription: '',
      errorAmount: '',
      errorDate: '',
      dateMessage: '',
    };
  }
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description, errorDescription: '' }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      return this.setState(() => ({ amount, errorAmount: '' }));
    }
    return undefined;
  };
  onDateChange = (createdAt) => { // this function takes a moment input
    this.setState(() => ({ createdAt, errorDate: '' }));
    if (!moment(createdAt)._isValid) {
      return this.setState(() => ({ dateMessage: 'DD/MM/YYYY' }));
    }
    return this.setState(() => ({ dateMessage: '' }));
  };
  onFocusChange = ({ focused }) => { // destructured somehow. not sure where '.focused' is getting pulled from. maybe the focused property in the <SingleDatePicker component?
    this.setState(() => ({ calendarFocused: focused }));
  };
  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount || !this.state.createdAt) {
      if (!this.state.description) {
        this.setState(() => ({ errorDescription: 'Description Required' }));
      }
      if (!this.state.amount) {
        this.setState(() => ({ errorAmount: 'Amount Required' }));
      }
      if (!this.state.createdAt) {
        this.setState(() => ({ errorDate: 'Dates must be formatted DD/MM/YYYY' }));
      }
      return;
    }
    this.props.onSubmit({
      description: this.state.description,
      amount: parseFloat(this.state.amount, 10) * 100, // turns the string into a real number (opposite is parseInt). second arg is the base 10, and the third is because we want cents
      createdAt: this.state.createdAt.valueOf(), // value of is a moment function that takes a moment object and returns it to a unix epock milliseconds value
      note: this.state.note,
    });
  }
  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.errorDescription && <p className="form__error">{this.state.errorDescription}</p>}
        {this.state.errorAmount && <p className="form__error">{this.state.errorAmount}</p>}
        {this.state.errorDate && <p className="form__error">{this.state.errorDate}</p>}
        <input
          type="text"
          placeholder="Description"
          autoFocus
          className="text-input"
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <input
          type="text"
          placeholder="Amount"
          className="text-input"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />
        <SingleDatePicker// date ondatechange focused and onfocuschange are all required props of the component. there are additional optional ones you can add too
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          // the below props are optional
          numberOfMonths={1}
          isOutsideRange={() => false}
        />{this.state.dateMessage && <p>{this.state.dateMessage}</p>}
        <textarea
          placeholder="Add a note for your expense (optional)"
          className="text-area"
          value={this.state.note}
          onChange={this.onNoteChange}
        />
        <div>
          <button className="button">{this.props.buttonContent ? this.props.buttonContent : 'Add Expense' }</button>
        </div>
      </form>
    );
  }
}
