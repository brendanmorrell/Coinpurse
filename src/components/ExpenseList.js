import React from 'react';
import { connect } from 'react-redux';

import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';
// adding 'export' is so you can grap the non-connected component for testing. the default export is the one you actually use for the app, and it is connected to the state via redux



export const ExpenseList = (props) => (
  <div>
    {
      props.expenses.length === 0 ? (
        <p>No Expenses currently</p>
      ) : (
        props.expenses.map((expense) => {
          return <ExpenseListItem key={expense.id} {...expense} />;
        })
      )
    }
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
    filters: state.filters,
  };
};

export default connect(mapStateToProps)(ExpenseList);// with connect, you put in the state info you want access to as the first argument, and this returns a function that you then input an argument in using () and insert the component you want to connect to the store.
// the syntax is a bit weird, but the component is placed outside the connect function
