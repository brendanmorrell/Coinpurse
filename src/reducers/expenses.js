//  REDUCERS
// 1. Reducers are pure functions (doesn't use or change anything from outside the function scope)
// 2. Never change state or action (you want to return a new state usually if that's what you end up trying to d0)


// Expenses Reducer
const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [
        ...state,
        action.expense,
      ];
    case 'REMOVE_EXPENSE':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates,
          };
        }
        return expense;
      });
    default: {
      return state;
    }
  }
};
