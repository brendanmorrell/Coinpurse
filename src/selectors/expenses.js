// Get Visible expenses
import moment from 'moment';


export default (expenses, {
  text,
  sortBy,
  startDate,
  endDate,
  sortReverse,
  minAmount,
  maxAmount,
}) => expenses.filter((expense) => {
  // odd thing that is necessary for handling zero in min and max
  if ((parseInt(maxAmount, 10) === 0 && expense.amount > 0) || (parseInt(minAmount, 10) === 0 && expense.amount < 0)) {
    return false;
  }
  const minAmountDollars = minAmount * 100;
  const maxAmountDollars = maxAmount * 100;
  const createdAtMoment = moment(expense.createdAt);
  const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
  const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
  const textMatch = expense.description.toLowerCase().includes(text.toLowerCase()) || expense.description.toLowerCase().includes(text.toLowerCase());
  const minAmountMatch = minAmountDollars ? expense.amount >= minAmountDollars : true;
  const maxAmountMatch = maxAmountDollars ? expense.amount <= maxAmountDollars : true;


  return startDateMatch && endDateMatch && textMatch && minAmountMatch && maxAmountMatch;
}).sort((a, b) => {
  if (!sortReverse) {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (a.amount > b.amount) { // SORT BY AMOUNT BELOW
      return -1;
    } else if (a.amount < b.amount) {
      return 1;
    }
    return a.description < b.description ? -1 : 1;// set equivalent values as alphabetical. could change this to defaulting to createdAt when amount is the same
  } else if (sortReverse) {
    if (sortBy === 'date') {
      return a.createdAt > b.createdAt ? 1 : -1;
    } else if (a.amount < b.amount) {
      return -1;
    } else if (a.amount > b.amount) {
      return 1;
    }
    return a.description > b.description ? -1 : 1;
  }
  return undefined;
});
