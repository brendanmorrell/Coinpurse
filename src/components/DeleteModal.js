import React from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';

import { startRemoveExpense } from '../actions/expenses';
import { toggleModal, pushToDashboard } from '../actions/modal';

export class DeleteModal extends React.Component {
  onRemoveExpense = () => {
    const { modal } = this.props;
    const { id } = this.props;
    this.props.startRemoveExpenseProp(id);
    this.props.toggleModal(!modal);
    this.props.pushToDashboard(true);
  };
  doNotRemoveExpense = () => {
    const { modal } = this.props;
    this.props.toggleModal(!modal);
  };
  render() {
    return (
      <div>
        <Modal
          isOpen={this.props.modal}
          className="modal"
          onAfterOpen={() => ({})}
          onRequestClost={() => ({})}
          closeTimeoutMS={11100}
          contentLabel="Modal"
        >
          <h3 className="modal__title">Are you sure you want to delete this expense?</h3>
          <div className="modal-div">
            <button className="button" onClick={this.onRemoveExpense}>Yes</button>
            <button className="button" onClick={this.doNotRemoveExpense}>No</button>
          </div>
        </Modal>
      </div>
    );
  }
}


const mapDispatchToProps = (dispatch, props) => ({
  startRemoveExpenseProp: (id) => dispatch(startRemoveExpense(id)),
  toggleModal: (modal) => dispatch(toggleModal(modal)),
  pushToDashboard: (pushTD) => dispatch(pushToDashboard(pushTD)),
});
const mapStateToProps = (state, props) => ({ modal: state.modal.modal });
export default connect(mapStateToProps, mapDispatchToProps)(DeleteModal);
