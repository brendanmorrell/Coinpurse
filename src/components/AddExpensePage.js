import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/dashboard');// this is an included action on react components. you can see it in the react devtools. PUSh takes a single argument which is the address you want to go to, just like links
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense)),
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);// mapDispatch to props allows us to put dispatch calls as props, thus allowing us to grab them as calls during testing as opposed
