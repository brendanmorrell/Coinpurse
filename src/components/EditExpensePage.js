import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';


export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);// here and below we can use props.match or props.expense since they are both being sent into the EditExpensePage component
    this. props.history.push('/dashboard');
  };
  onRemove = () => {
    this.props.startRemoveExpense(this.props.expense.id);
    this.props.history.push('/dashboard');
  };
  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
          buttonContent="Submit Edit"
        />
        <button
          onClick={this.onRemove}
        >Remove
        </button>
      </div>
    );
  };
};

const mapStateToProps = (state, props) => ({expense: state.expenses.find((expense) => expense.id === props.match.params.id)});

const mapDispatchToProps = (dispatch, props) => {
  return {
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
