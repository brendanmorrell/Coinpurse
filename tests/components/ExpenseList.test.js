import React from 'react';
import { shallow } from 'enzyme';

import { ExpenseList } from '../../components/ExpenseList';
// ^you have to change the component to a non-default export so you can grab the non-connected (no state/props so you can pass whatever you want in here for testing) component, which will be grabbed here, and then when you import it as a default (for the real app) you are importing the connected one with the state and props mapped to it
import expenses from '../fixtures/expenses'; // this is the test expenses array we made for all our other tests




test('Should render the ExpenseList with Expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
});



test(`Should render <p>No Expenses Currently</p> when there are no expenses`, () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />);
  expect(wrapper).toMatchSnapshot();
});
