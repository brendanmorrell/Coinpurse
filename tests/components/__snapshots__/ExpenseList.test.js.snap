// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render <p>No Expenses Currently</p> when there are no expenses 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item--message"
    >
      <span>
        No expenses currently
      </span>
    </div>
  </div>
</div>
`;

exports[`Should render the ExpenseList with Expenses 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <Connect(ExpenseListItem)
      amount={195}
      createdAt={0}
      description="Gum"
      id="1"
      key="1"
      note=""
    />
    <Connect(ExpenseListItem)
      amount={109500}
      createdAt={-172800000}
      description="REnt"
      id="2"
      key="2"
      note=""
    />
    <Connect(ExpenseListItem)
      amount={4500}
      createdAt={172800000}
      description="credit card"
      id="3"
      key="3"
      note=""
    />
  </div>
</div>
`;
